### PR TITLE
[Security] fix Check if it has session before getSession()

### LINKED ISSUE #42258: [SecurityHttp] if  don't have session,logout will throw SessionNotFoundException
**Symfony version(s) affected**: 5.2.4

**Description**
If not support of sessions in the app，`getSession()`wile  throw SessionNotFoundException in the `onLogout()`

```php
class SessionLogoutListener implements EventSubscriberInterface
{
    public function onLogout(LogoutEvent $event): void
    {
        $event->getRequest()->getSession()->invalidate();
    }

    public static function getSubscribedEvents(): array
    {
        return [
            LogoutEvent::class => 'onLogout',
        ];
    }
}
```

**Possible Solution**

Check if it has session before `getSession()`

### PR DESCRIPTION
**Pull request by mousezheng**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #42258 
| License       | MIT

Check if it has session before getSession(),for details, see [issues#42258](https://github.com/symfony/symfony/issues/42258)

### CODE REVIEW

**Comment by StaffNowa** [comment]
I think you will need rebase to 5.3

```
This version is no longer maintained. Consider upgrading to Symfony 5.3.
```

**Comment by derrabus** [comment]
You've submitted your patch against the 5.4 branch although 5.2 should be the correct target. Can you please rebase your changes?

Furthermore, can you please add a test covering your changes?

**Comment by mousezheng** [comment]
@derrabus 

The 5.2 version is no longer maintained. So consider upgrading to Symfony 5.3. 
I submitted my patch against the 5.3 branch.

And , I already add a test covering my changes. Please advise thanks.

**Comment by chalasr** [comment]
Don't this apply to `SessionLogoutHandler` as well?

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+class SessionLogoutListenerTest extends TestCase
+{
+    public function testOnLogoutIfHasNotSession()
+    {
+        $session = $this->createMock(Session::class);
+        $session->expects($this->never())->method('invalidate');
+
+        $request = $this->createMock(Request::class);
+        $request->expects($this->once())->method('hasSession')->willReturn(false);
```
Comment (derrabus) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]:
```suggestion
        $request->method('hasSession')->willReturn(false);
        $request->expects($this->never())->method('getSession');
```

I don't think it's important how often `hasSession()` is called, but let's make sure `getSession()` is not called.

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function testOnLogoutIfHasSession()
+    {
+        $session = $this->createMock(Session::class);
+        $session->expects($this->once())->method('invalidate');
+
+        $request = $this->createMock(Request::class);
+        $request->expects($this->once())->method('getSession')->willReturn($session);
+        $request->expects($this->once())->method('hasSession')->willReturn(true);
```
Comment (derrabus) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]:
```suggestion
        $request->method('getSession')->willReturn($session);
        $request->method('hasSession')->willReturn(true);
```

Again, it should not matter how often those methods are called. The important assertion is that `$session->invalidate()` is called.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
> The 5.2 version is no longer maintained.

Yeah, my comment did not age well. 😅 We've just release the final 5.2 release, so 5.3 is the correct target now.

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+use Symfony\Component\Security\Http\EventListener\SessionLogoutListener;
+
+class SessionLogoutListenerTest extends TestCase
+{
+    public function testOnLogoutIfHasNotSession()
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]:
```suggestion
    public function testOnLogoutIfHasNoSession()
```

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Symfony\Component\Security\Http\Tests\EventListener;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+use Symfony\Component\Security\Http\EventListener\SessionLogoutListener;
+
+class SessionLogoutListenerTest extends TestCase
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php:20]:
```suggestion
final class SessionLogoutListenerTest extends TestCase
```

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php (hunk 1 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]: Should be `$request->expects($this->any())->method('hasSession')->willReturn(false);` in that case, right?

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+{
+    public function testOnLogoutIfHasNotSession()
+    {
+        $session = $this->createMock(Session::class);
+        $session->expects($this->never())->method('invalidate');
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]:
```suggestion
```

(this tests when the session has not been set, so it's not used in this test)

**Comment by wouterj** [comment]
Hi @mousezheng! Thanks for preparing this PR so quickly and even writing a test case.

There are some mostly aesthetic suggestions for your test. Don't worry about too much about them, each library has its own unwritten testing practices :)

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php]: @wouterj That's the same, just with more characters. 😉

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php (hunk 4 above)
Comment (mousezheng) [reply src/Symfony/Component/Security/Http/Tests/EventListener/SessionLogoutListenerTest.php:20]: The test class don't need to be `final`. Here, be consistent with others.

**Comment by mousezheng** [comment]
> Don't this apply to `SessionLogoutHandler` as well?

@chalasr  `SessionLogoutHandler` deprecated since Symfony 5.1.  in 4.4 this may be a problem.

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
> @chalasr `SessionLogoutHandler` deprecated since Symfony 5.1.

But that does not mean that we can leave it broken. If it needs to be fixed, we should fix it.

**Review by wouterj — APPROVED** [review]

**Comment by wouterj** [comment]
> But that does not mean that we can leave it broken. If it needs to be fixed, we should fix it.

But we can't include the fix in this PR, as that fix should target 4.4 and not 5.3 :)

**Comment by derrabus** [comment]
Thank you @mousezheng.

**Pull request merged**